indico/modules/events/management/controllers/protection.py

```diff
@@ -63,42 +70,86 @@ def _process(self):
                                 endpoint='event_management.acl')
 
 
-class RHEventProtection(RHManageEventBase):
+class RHEventProtectionBase(RHManageEventBase):
+    def __init__(self):
+        super(RHEventProtectionBase, self).__init__()
+        selectable_permissions = {k: v for k, v in get_available_permissions(Event).viewitems() if v.user_selectable}
+        self.whitelisted_permissions = set(selectable_permissions.keys()) | {'edit', 'access'}
+        self.permissions_tree = {
+            'edit': {
+                'title': _('Event Management'),
+                'children': {
+                    v.name: {'title': v.friendly_name} for k, v in selectable_permissions.viewitems()
+                }
+            }
+        }
+        self.permissions_tree['edit']['children']['access'] = {'title': _('Access')}
+        self.permissions_tree['edit']['children'] = OrderedDict(
+            sorted(self.permissions_tree['edit']['children'].items(), key=lambda t: t[0])
+        )
+
+
+class RHEventProtection(RHEventProtectionBase):
     """Show event protection"""
 
     NOT_SANITIZED_FIELDS = {'access_key'}
 
     def _process(self):
         form = EventProtectionForm(obj=FormDefaults(**self._get_defaults()), event=self.event)
         if form.validate_on_submit():
+            permission_principals = defaultdict(set)
+            for principal_data in form.permissions.data:
+                for permission in principal_data[1]:
+                    permission_principals[permission].add(principal_from_fossil(principal_data[0], allow_emails=True,
+                                                                                allow_networks=True))
+            for permission in self.whitelisted_permissions:
+                if permission == 'edit':
+                    update_object_principals(self.event, permission_principals.get(permission, set()), full_access=True)
```

Comment:
Don't call `update_object_principals` in a loop. This spams the event log with one entry per call! Instead, build the args (`full_access`, `read_access`, `permissions`) and then call it once. Since `update_object_principals` only supports updating one permission/access bit at a time, you might want to simply add another function that deals with the data you have here, so you'd just call something like `update_permissions(self.event, permission_principals, available_permissions)`

Reply:
I don't understand how I can do it in one single call. As you said, `update_object_principals` only supports updating one permission/access bit at a time, so how can I update all the permissions (since all of than might have changed when saving) in a unique call? Don't you need one call per permission (what I'm doing now)?

Reply:
There is `event.update_principal()` where you pass the principal (user/group/etc.) and the list of roles to add/remove/set and the read/manage privs. You can call this directly, so it's one call per user no matter how many permissions.

Reply:
Deleted entries need to be handled separately (like we probably do right now); either by removing all privileges via `update_principal` or by calling `delete_principal`